### PR TITLE
Workflow graph structure

### PR DESCRIPTION
**Pull request by antho1404**
Update the structure of the workflow with a graph approach.

The workflow definition is now created based on a list of nodes and a list of edges that connects these nodes. 
Every node need to be unique in the workflow and will be identifiable. With the node identifiable we can now store the current step when creating execution and can drop the height support that was not valid when a workflow was triggered by a result of the same workflow and the same task.

With this new `stepID` we can simply find the children of this node in the workflow definition graph and execute all the children's tasks.

- SDK of execution updated to accept the `stepID`
- API updated with the new workflow struct and execution stepID

## TO TEST

Create and start the services with the cli or API.

```json
// ServiceA
{"name":"serviceA","tasks":[{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskX"},{"key":"emits"}],"events":[{"data":[{"type":"String","key":"message"}],"key":"eventX"}],"configuration":{},"hash":"3kWX7CYtNwbeb1h35HiNAPctnh94MU8zstT7PSuusxyy","sid":"serviceA","source":"QmUojyMjo2JUvhktLU4royUBDEsVzU4z1onPovGpVQAFDW"}

// ServiceB
{"name":"serviceB","tasks":[{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskY"}],"configuration":{},"hash":"8ExNFweHXavpZSE9iTwjBFSDWzKcRpCXEc4BuX1UwFcg","sid":"serviceB","source":"Qmf2e92jNxxmYyjxEFrCBJdCNUS1LKNWet2ZQt3ByEUpAP"}

// ServiceC
{"name":"serviceC","tasks":[{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskZ"},{"inputs":[{"type":"String","key":"message"}],"outputs":[{"type":"String","key":"message"}],"key":"taskA"}],"configuration":{},"hash":"qTtisJG3Bso5rRxSHhnrQSkJpC3JGePD3ovvuEGVgeg","sid":"serviceC","source":"QmPuHX3LumnDkrxY3wUR8J2J9xmXGw8yTVy5jf7P24MprN"}
```

Create one or all the following workflow(s) with the API 

eg: `grpcurl -plaintext -d "$(cat result.json| jq -c .)" localhost:50052 api.Workflow/Create`

```json
// result.json
{"key":"result","trigger":{"type":2,"instanceHash":"A7712VMAiT2ehXV1nHHUnkvyTbZKVHeFxBZZj4Gx7MZk","key":"taskX","initilNode":"task1"},"nodes":[{"key":"task1","instanceHash":"DUto1J3ZAheVL6PGrwYTtGF4Bnw15CQocVdEmqkHSGbB","taskKey":"taskY"}, {"key":"task2","instanceHash":"FjFzXXme2dkLBsgPk2xgyiPD6dX5EN1dBLGqW6G9tStf","taskKey":"taskZ"}],"edges":[{"src":"task1","dst":"task2"}]}

// event.json
{"key":"event","trigger":{"type":1,"instanceHash":"A7712VMAiT2ehXV1nHHUnkvyTbZKVHeFxBZZj4Gx7MZk","key":"eventX","initialNode":"task1"},"nodes":[{"key":"task1","instanceHash":"DUto1J3ZAheVL6PGrwYTtGF4Bnw15CQocVdEmqkHSGbB","taskKey":"taskY"},{"key":"task2","instanceHash":"FjFzXXme2dkLBsgPk2xgyiPD6dX5EN1dBLGqW6G9tStf","taskKey":"taskZ"}],"edges":[{"src":"task1","dst":"task2"}]}

// result-from-other-workflow
{"key":"result-from-other-workflow","trigger":{"type":2,"instanceHash":"DUto1J3ZAheVL6PGrwYTtGF4Bnw15CQocVdEmqkHSGbB","key":"taskY","initialNode":"task1"},"nodes":[{"key":"task1","instanceHash":"FjFzXXme2dkLBsgPk2xgyiPD6dX5EN1dBLGqW6G9tStf","taskKey":"taskZ"}],"edges":[]}
```

Then you see the workflow by running either an event with 
```bash
mesg-cli service:execute serviceA emits
```
or a task
```bash
mesg-cli service:execute serviceA taskX --data message=hello
```

### CODE REVIEW

File: protobuf/types/execution.proto [hunk 1]
```diff
@@ -59,4 +56,10 @@ message Execution {
 
   // tags are optionally associated with execution by the user.
   repeated string tags = 10;
+
+  // wokflowHash is the unique hash of the workflow associated to this execution.
+  string workflowHash = 11;
+
+  // step in the workflow
```
Comment (NicolasMahe) [inline protobuf/types/execution.proto]:
```suggestion
  // step of the workflow.
```

File: protobuf/types/execution.proto [hunk 2]
```diff
@@ -59,4 +56,10 @@ message Execution {
 
   // tags are optionally associated with execution by the user.
   repeated string tags = 10;
+
+  // wokflowHash is the unique hash of the workflow associated to this execution.
```
Comment (NicolasMahe) [inline protobuf/types/execution.proto]:
```suggestion
  // workflowHash is the unique hash of the workflow associated to this execution.
```

File: protobuf/types/workflow.proto [hunk 3]
```diff
@@ -30,20 +30,28 @@ message Workflow {
       string value = 3;         // Value of the filter
     }
 
-    Type type = 1;         // Type of trigger.
+    Type type = 1;                // Type of trigger.
     string instanceHash = 2;      // Hash of the instance that triggers the workflow.
     string key = 3;               // Key that triggers the workflow (event key or task key).
     repeated Filter filters = 4;  // List of filters to apply on the data of the event/result.
+    string initialNode = 5;       // First node to trigger when the workflow starts
```
Comment (NicolasMahe) [inline protobuf/types/workflow.proto]:
```suggestion
    string nodeKey = 5;       // First node to trigger when the workflow starts.
```

File: protobuf/types/workflow.proto [hunk 4]
```diff
@@ -30,20 +30,28 @@ message Workflow {
...
   }
 
-  // Definition of the task to execute when the workflow is triggered.
-  message Task {
-    string instanceHash = 1;  // Hash of the instance to execute.
-    string taskKey = 2;       // Task of the instance to execute.
+  // Definition of the node to execute when the workflow is triggered.
+  message Node {
+    string key = 1;           // Key that identify the node
```
Comment (NicolasMahe) [inline protobuf/types/workflow.proto]:
```suggestion
    string key = 1;           // Key that identifies the node.
```

File: protobuf/types/workflow.proto [hunk 5]
```diff
@@ -30,20 +30,28 @@ message Workflow {
...
+    string instanceHash = 2;  // Hash of the instance to execute.
+    string taskKey = 3;       // Task of the instance to execute.
+  }
+
+  message Edge {
+    string src = 1; // Source of the edge
```
Comment (NicolasMahe) [inline protobuf/types/workflow.proto]:
```suggestion
    string src = 1; // Source of the edge.
```

File: protobuf/types/workflow.proto [hunk 6]
```diff
@@ -30,20 +30,28 @@ message Workflow {
...
+    string dst = 2; // Destination of the edge
```
Comment (NicolasMahe) [inline protobuf/types/workflow.proto]:
```suggestion
    string dst = 2; // Destination of the edge.
```

File: workflow/type.go [hunk 7]
```diff
@@ -22,23 +22,32 @@ const (
...
+	Nodes   []Node    `hash:"name:3" validate:"dive,required"`
+	Edges   []Edge    `hash:"name:4" validate:"dive,required"`
 }
 
-// Task describes the instructions for the workflow to execute a task
-type Task struct {
+// Node describes the instructions for the workflow to execute a task
+type Node struct {
+	Key          string    `hash:"name:1" validate:"required"`
 	InstanceHash hash.Hash `hash:"name:1" validate:"required"`
```
Comment (NicolasMahe) [inline workflow/type.go]:
```suggestion
	InstanceHash hash.Hash `hash:"name:2" validate:"required"`
```

File: workflow/type.go [hunk 8]
```diff
@@ -22,23 +22,32 @@ const (
...
-	TaskKey      string    `hash:"name:2" validate:"printascii"`
+	TaskKey      string    `hash:"name:2" validate:"required,printascii"`
```
Comment (NicolasMahe) [inline workflow/type.go]:
```suggestion
	TaskKey      string    `hash:"name:3" validate:"required,printascii"`
```

File: workflow/type.go [hunk 9]
```diff
@@ -22,23 +22,32 @@ const (
...
 }
 
 // Trigger is an event that triggers a workflow
 type Trigger struct {
 	InstanceHash hash.Hash        `hash:"name:1" validate:"required"`
-	Key          string           `hash:"name:2" validate:"printascii"`
+	Key          string           `hash:"name:2" validate:"required,printascii"`
 	Type         TriggerType      `hash:"name:3" validate:"required"`
 	Filters      []*TriggerFilter `hash:"name:4" validate:"dive,required"`
+	InitialNode  string           `hash:"name:5" validate:"required"`
```
Comment (NicolasMahe) [inline workflow/type.go]:
```suggestion
	NodeKey  string           `hash:"name:5" validate:"required"`
```

File: workflow/type.go [hunk 10]
```diff
@@ -22,23 +22,32 @@ const (
 // Workflow describes a workflow of a service
 type Workflow struct {
 	Hash    hash.Hash `hash:"-" validate:"required"`
-	Trigger *Trigger  `hash:"name:1" validate:"required"`
-	Tasks   []*Task   `hash:"name:2" validate:"required"`
-	Key     string    `hash:"name:3" validate:"required"`
+	Key     string    `hash:"name:1" validate:"required"`
+	Trigger Trigger   `hash:"name:2" validate:"required"`
+	Nodes   []Node    `hash:"name:3" validate:"dive,required"`
```
Comment (krhubert) [inline workflow/type.go]: What do you think about moving the graph as a separate structure? In that case you can control and test all graph functions like searching path, walking order and more without Workflow and workflow will just use it like:

```go
type Workflow struct {
    // ....
    g wgraph
}
```

File: workflow/type.go [hunk 11]
```diff
@@ -22,23 +22,32 @@ const (
...
+// Node describes the instructions for the workflow to execute a task
+type Node struct {
+	Key          string    `hash:"name:1" validate:"required"`
+	InstanceHash hash.Hash `hash:"name:2" validate:"required"`
+	TaskKey      string    `hash:"name:3" validate:"required,printascii"`
+}
+
+// Edge describes the instructions for the workflow to execute a task
+type Edge struct {
+	Src string `hash:"name:1" validate:"required"`
```
Comment (krhubert) [inline workflow/type.go]: I don't see a validation if the edge has proper src and dst.

Comment (antho1404) [reply workflow/type.go]: yes not yet, I wanted to add the validation of nodes/edges and also the instances on another pr

File: workflow/type.go (hunk 10 above)
Comment (antho1404) [reply workflow/type.go]: for now, the workflow is pretty much only the graph but yes I agree that could be nice. I would still keep that as a future refactoring if the package gets bigger or we need to reuse the graph

**Comment by NicolasMahe** [comment]
Fixed workflows to send to API:
```
// result.json
{"key":"result","trigger":{"type":2,"instanceHash":"HX6xrMxEZRgSN97qL6GUo6aAqpnrNDWoPiS1wae9Qi2s","key":"taskX","nodeKey":"task1"},"nodes":[{"key":"task1","instanceHash":"37Hmvhp2xmH9JPmipk1ES2ATbSVJPGt7z2cD3VDJ1dtf","taskKey":"taskY"}, {"key":"task2","instanceHash":"GbDa3xsmzY3V6HvF8XKL9gxvV6aFZoj5nNvCrFq8GDTd","taskKey":"taskZ"}],"edges":[{"src":"task1","dst":"task2"}]}

// event.json
{"key":"event","trigger":{"type":1,"instanceHash":"HX6xrMxEZRgSN97qL6GUo6aAqpnrNDWoPiS1wae9Qi2s","key":"eventX","nodeKey":"task1"},"nodes":[{"key":"task1","instanceHash":"37Hmvhp2xmH9JPmipk1ES2ATbSVJPGt7z2cD3VDJ1dtf","taskKey":"taskY"},{"key":"task2","instanceHash":"GbDa3xsmzY3V6HvF8XKL9gxvV6aFZoj5nNvCrFq8GDTd","taskKey":"taskZ"}],"edges":[{"src":"task1","dst":"task2"}]}

// result-from-other-workflow
{"key":"result-from-other-workflow","trigger":{"type":2,"instanceHash":"37Hmvhp2xmH9JPmipk1ES2ATbSVJPGt7z2cD3VDJ1dtf","key":"taskY","nodeKey":"task1"},"nodes":[{"key":"task1","instanceHash":"GbDa3xsmzY3V6HvF8XKL9gxvV6aFZoj5nNvCrFq8GDTd","taskKey":"taskZ"}],"edges":[]}
```

**Pull request merged**